lib/Transforms/Secretize/WrapGeneric.cpp

```diff
@@ -0,0 +1,93 @@
+#include <iostream>
+
+#include "include/Dialect/Secret/IR/SecretDialect.h"
+#include "include/Dialect/Secret/IR/SecretOps.h"
+#include "include/Dialect/Secret/IR/SecretTypes.h"
+#include "include/Transforms/Secretize/Passes.h"
+#include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/IRMapping.h"             // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"  // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_WRAPGENERIC
+#include "include/Transforms/Secretize/Passes.h.inc"
+
+struct WrapWithGeneric : public OpRewritePattern<func::FuncOp> {
+  WrapWithGeneric(mlir::MLIRContext *context)
+      : mlir::OpRewritePattern<func::FuncOp>(context) {}
+
+  LogicalResult matchAndRewrite(func::FuncOp op,
+                                PatternRewriter &rewriter) const override {
+    bool hasSecrets = false;
+
+    SmallVector<Type, 4> newInputs;
+    for (unsigned i = 0; i < op.getNumArguments(); i++) {
+      if (op.getArgAttr(i, secret::SecretDialect::kArgSecretAttrName) !=
+          nullptr) {
+        hasSecrets = true;
+        op.removeArgAttr(i, secret::SecretDialect::kArgSecretAttrName);
+
+        auto newTy = secret::SecretType::get(op.getArgument(i).getType());
+        op.getArgument(i).setType(newTy);  // Updates the block argument type.
+        newInputs.push_back(newTy);
+      }
+    }
+
+    if (!hasSecrets) {
+      // Match failure, no secret inputs.
+      return failure();
+    }
+
+    llvm::SmallVector<Type, 4> newOutputs;
```

Comment:
nit, you might appreciate `llvm::map_range` here:

```suggestion
    auto newOutputs = llvm::map_range(
        op.getResultTypes(), [](Type t) { return secret::SecretType::get(t); });

    op.setFunctionType(
        FunctionType::get(getContext(), {newInputs}, TypeRange{newOutputs}));
```